### ISSUE TITLE
Information for using Start-Parameters recalc/compact

### ISSUE THREAD

**Issue by ReinerWatzlawik** [jokay/docker-ccu-historian#140]
Hello,
first I have to say this work is great. Thanks a lot for this! 
After changing the configuration of many data points in ccu-historian into the docker container I want to recalc and compact the database.
In the past - without docker container - i have done this manually into a terminal session, so I know this process.

When I stop the container it is not possible to get access to start ccu-historian with start paramters.
Could you give a short explanation the process for users (I'm not an expert...) how to do this?

Many thanks!

**Comment by jokay** [comment]
The environment variable [CONFIG_KEEP_MONTHS](https://github.com/jokay/docker-ccu-historian#configuration) may be what you are looking for.

It includes the maintenance steps for `-clean`, `-recalc` and `-compact`.

**Comment by ReinerWatzlawik** [comment]
Thank you for the fast answer! I didn't saw this in the documentation.

**Comment by PaulMerk** [comment]
Is there another possibility ? I do not want to delete data.
Thanks
Paul

**Comment by ReinerWatzlawik** [comment]
I set 36 month - think it's a long time. At CCU-Historian restart I saw in the log some datapoints with older data which are cleaned. 
You can test longer time - clean/recalc/compact should be started by each restart of the container.

**Comment by ReinerWatzlawik** [comment]
When the value of month will set to a (very) large number nothing will be cleaned. 
Only by recalculating single values will be dropped.

The (only?) disadvantage of this method is the noticeable longer time for starting ccu-historian. In my case it takes 3 Minutes (version 2.8.1 with 1070 data points an 860.000 entries on Synolgy DS218+). Normaly it takes 13 seconds. 
But since you don't have to restart often anyway, that shouldn't be a problem.

@jokay: could you add to the documenation? When no number is given by CONFIG_KEEP_MONTHS the environment variable has no effect... So it easily could used for manual startet maintanance.

**Comment by jokay** [comment]
Yes, if the environment variable is not set, it's not taken into consideration :wink:

I will add another environment variable only for the maintenance which can then be used in a future release.

**Comment by ReinerWatzlawik** [comment]
possible to set to 'once' (with automatic reset to'blank') or e.g. monthly (without reset)?

**Comment by jokay** [comment]
Not sure how you think I will be able to reset an environment variable defined outside the container?

**Comment by jokay** [comment]
> @jokay: could you add to the documenation? When no number is given by CONFIG_KEEP_MONTHS the environment variable has no effect... So it easily could used for manual startet maintanance.

I don't think there needs to be any documentation for this if you are going to use a config param without a value.

**Comment by PaulMerk** [comment]
> I will add another environment variable only for the maintenance which can then be used in a future release.

good idea !!!
users haveto be patient, the clean … process may run for minutes, even hours depending on the size of the database

**Comment by jokay** [comment]
@ReinerWatzlawik, @PaulMerk could you please tell me your use case you would like to be covered?

Would it work for you to have an environment variable, e.g. `CONFIG_MAINTENANCE` *(type boolean)* which can be set to **true** to execute `-recalc` followed by `-compact` and prior starting CCU-Historian?

**Comment by ReinerWatzlawik** [comment]
I think about the use case:
Sometimes I create new sensors and don't know usefull parameters für preprocessing. 
So I let it run and look after some time to the number of database entries for the new data points.
(Or I forget to set the parameters. ) 
After setting (or adjusting)  the preprocessing parameters it would make sense to recalc and compact the database. 
In this case it would be ok to set an environment variable . 
Recalc/compact does not have to be executed immediately but at a defined time.

I don't know what are the possibilties: is it possible to reset the environment variable to start it only once? 
E.g. I set it 'true', recalc/compact will be done in the night (e.g. 23:00 pm), and reset the environment variable to 'false'?

Or is it possible (less time consuming?) to set daily or weekly or monthly (that's my favorit) for regular automatic maintenance?

Sorry, I don't know the possibilities. In the end, a simple, possibly regular, automatic maintenance of the database would be a great thing!

**Comment by jokay** [comment]
@ReinerWatzlawik it seems similar to https://github.com/jokay/docker-ccu-historian/issues/10 regarding the scheduled maintenance.

It makes no sense to reset environment variables as they would be set to their initial value on (re-)starting the container.

Cron job handling is not setup in the container but I can provide a specific startup handling based on environment params.

**Comment by ReinerWatzlawik** [comment]
is it possible to restart the container bei task-managment ('Aufgabenplaner' in DSM)?

**Comment by jokay** [comment]
> I propose an environment variable "compact database at next start only: YES", which would be reset to "NO" for the next start of the container.

@PaulMerk I don't think this is possible. I can set it to "NO" in the container after the first run but it will be reset to "YES" on the next start. There is no persistence for environment variables if not saved somehow to a persistent storage.

**Comment by jokay** [comment]
I don't think it is a good idea to insert config params for the container which are unrelated to the CCU-Historian itself.

But still I can provide an additional environment variable `CONFIG_MAINTENANCE` which can be set to **true/false** anytime and which will be taken into consideration on each start of the container.

**Comment by PaulMerk** [comment]
perfect !

**Comment by jokay** [comment]
The next release will include the environment variable `CONFIG_MAINTENANCE`, see [here](https://github.com/jokay/docker-ccu-historian#configuration), which gives you the following options:

#### CONFIG_KEEP_MONTHS=12

> *(CONFIG_MAINTENANCE **not** set)*

Executes **-clean**, **-recalc** and **-compact** before starting.

```sh
2022-01-12 20:18:42|INFO   |jokay/ccu-historian 2.8.3
2022-01-12 20:18:42|INFO   |Creating config file ...
2022-01-12 20:18:42|INFO   |Running database maintenance 'clean' (removes all data before 2021-01-13) ...
2022-01-12 20:18:43|INFO   |CCU-Historian V2.8.3
2022-01-12 20:18:43|INFO   |(C)MDZ (info@ccu-historian.de)
2022-01-12 20:18:44|INFO   |Starting base services
2022-01-12 20:18:44|INFO   |Connecting to database
2022-01-12 20:18:44|INFO   |Starting database web server
2022-01-12 20:18:45|INFO   |Starting cleaning of time series date before Wed Jan 13 00:00:00 CET 2021
2022-01-12 20:18:45|INFO   |Summary: 0 entries from 0 dropped (0.0 %); 0 seconds
2022-01-12 20:18:45|INFO   |Cleaning completed
2022-01-12 20:18:45|INFO   |Running database maintenance 'recalc' ...
2022-01-12 20:18:46|INFO   |CCU-Historian V2.8.3
2022-01-12 20:18:46|INFO   |(C)MDZ (info@ccu-historian.de)
2022-01-12 20:18:47|INFO   |Starting base services
2022-01-12 20:18:47|INFO   |Connecting to database
2022-01-12 20:18:47|INFO   |Starting database web server
2022-01-12 20:18:47|INFO   |Starting recalculation of compressed data points
2022-01-12 20:18:48|INFO   |Summary: 0 entries from 0 dropped (0.0 %); 0 seconds
2022-01-12 20:18:48|INFO   |Recalculation completed
2022-01-12 20:18:48|INFO   |Running database maintenance 'compact' ...
2022-01-12 20:18:49|INFO   |CCU-Historian V2.8.3
2022-01-12 20:18:49|INFO   |(C)MDZ (info@ccu-historian.de)
2022-01-12 20:18:50|INFO   |Starting base services
2022-01-12 20:18:50|INFO   |Starting compaction of database
2022-01-12 20:18:50|INFO   |Dumping database to temp.sql
2022-01-12 20:18:51|INFO   |Restoring database from temp.sql
2022-01-12 20:18:51|INFO   |Compaction of database completed
2022-01-12 20:18:51|INFO   |Starting CCU-Historian using the following config:
```

#### CONFIG_MAINTENANCE=true

> *(CONFIG_KEEP_MONTHS **not** set)*

Executes **-recalc** and **-compact** before starting.

```sh
2022-01-12 20:20:23|INFO   |jokay/ccu-historian 2.8.3
2022-01-12 20:20:23|INFO   |Running database maintenance 'recalc' ...
2022-01-12 20:20:24|INFO   |CCU-Historian V2.8.3
2022-01-12 20:20:24|INFO   |(C)MDZ (info@ccu-historian.de)
2022-01-12 20:20:25|INFO   |Starting base services
2022-01-12 20:20:25|INFO   |Connecting to database
2022-01-12 20:20:25|INFO   |Starting database web server
2022-01-12 20:20:26|INFO   |Starting recalculation of compressed data points
2022-01-12 20:20:26|INFO   |Summary: 0 entries from 0 dropped (0.0 %); 0 seconds
2022-01-12 20:20:26|INFO   |Recalculation completed
2022-01-12 20:20:26|INFO   |Running database maintenance 'compact' ...
2022-01-12 20:20:27|INFO   |CCU-Historian V2.8.3
2022-01-12 20:20:27|INFO   |(C)MDZ (info@ccu-historian.de)
2022-01-12 20:20:28|INFO   |Starting base services
2022-01-12 20:20:28|INFO   |Starting compaction of database
2022-01-12 20:20:28|INFO   |Dumping database to temp.sql
2022-01-12 20:20:29|INFO   |Restoring database from temp.sql
2022-01-12 20:20:29|INFO   |Compaction of database completed
2022-01-12 20:20:29|INFO   |Starting CCU-Historian using the following config:
```

#### CONFIG_MAINTENANCE=false

> *(CONFIG_KEEP_MONTHS **not** set)*

Executes nothing before starting.

```sh
2022-01-12 20:21:07|INFO   |jokay/ccu-historian 2.8.3
2022-01-12 20:21:07|INFO   |Starting CCU-Historian using the following config:
```

**Comment by ReinerWatzlawik** [comment]
Okay , I don't want to make it unnecessarily complicated. With these options I can solve the maintenance tasks.

But one idea or question still remains: is it possible to use the task-planer from synology DSM to start the container with different values of environment variables? 
I can stop/start (restart) the container with a script in a scheduled task ("docker container stop container-name" or "docker container start container-name"). 
If I know the syntax/possible params für starting these script it would be possible without manully action...

**Comment by jokay** [comment]
> is it possible to use the task-planer from synology DSM to start the container with different values of environment variables?

I can't help you with that Synology specific setup. Personally I use a Linux host with Docker daemon for my Docker containers.

May be someone can help on the [Synology Forum](https://community.synology.com/enu)?

**Comment by jokay** [comment]
Support for `CONFIG_MAINTENANCE` is merged and will be available in the next release, see [here](https://github.com/jokay/docker-ccu-historian/blob/main/CHANGELOG.md).

**Comment by jokay** [comment]
Feature is included in todays [2.9.0](https://github.com/jokay/docker-ccu-historian/releases/tag/2.9.0) release.
